executor/executor.go

```diff
@@ -1656,14 +1656,31 @@ func (e *UnionExec) Close() error {
 // Before every execution, we must clear statement context.
 func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 	vars := ctx.GetSessionVars()
-	sc := &stmtctx.StatementContext{
-		TimeZone:      vars.Location(),
-		MemTracker:    memory.NewTracker(memory.LabelForSQLText, vars.MemQuotaQuery),
-		DiskTracker:   disk.NewTracker(memory.LabelForSQLText, -1),
-		TaskID:        stmtctx.AllocateTaskID(),
-		CTEStorageMap: map[int]*CTEStorages{},
-		IsStaleness:   false,
+	// Save some variable in StmtCtx before reset it.
+	saveWarnings := vars.StmtCtx.GetWarnings()
+	saveAffectedRows := vars.StmtCtx.AffectedRows()
+	saveLastInsertID := vars.StmtCtx.LastInsertID
+	savePrevLastInsertID := vars.StmtCtx.PrevLastInsertID
+	saveErrCount, saveWarnCount := vars.StmtCtx.NumErrorWarnings()
+	saveInUpdateStmt, saveInDeleteStmt, saveInInsertStmt, saveInSelectStmt := vars.StmtCtx.InUpdateStmt, vars.StmtCtx.InDeleteStmt, vars.StmtCtx.InInsertStmt, vars.StmtCtx.InSelectStmt
+
+	if vars.TxnCtx.CouldRetry {
+		// Must construct new statement context object, the retry history need context for every statement.
+		// TODO: Maybe one day we can get rid of transaction retry, then this logic can be deleted.
+		vars.StmtCtx = &stmtctx.StatementContext{}
+	} else {
+		// Reuse mem to reduce object allocation.
+		vars.InitStatementContext()
 	}
+
+	sc := vars.StmtCtx
```

Comment:
Will `sc = stmtctx.StatementContext{}` do the trick, I mean, reset the sc? I don't understand why tests need to be modified if logic is same. One thing I noticed is that, `&xxx{}` will init all fields into the default zero value. But the new logic only init fields that is manually inited, other fields are not reseted.

Reply:
> I don't understand why tests need to be modified

+1.

Reply:
> Will sc = stmtctx.StatementContext{} do the trick

Yes, we can. 
But there is a drawback, when the struct are nested, the allocation still can't be avoided.

```
type SessionVars struct {
    StmtCtx *stmtctx.StatementContext
}

type StatementContext struct {
    RuntimeStatsColl *execdetails.RuntimeStatsColl
}

*vars.StmtCtx = stmtctx.StatementContext{}  // reset the data and reuse memory

vars.StmtCtx.RuntimeStatsColl = NewRuntimeStatsColl() // after reset, it's gone ... can't avoid allocating a new one!
```

The method adopted in this PR can avoid allocation totally.

```
type SessionVars struct {
    StmtCtx *stmtctx.StatementContext // point to the cached data
    cache struct {
        stmtctx.StatementContext
    }
}

type StatementContext struct {
    RuntimeStatsColl *execdetails.RuntimeStatsColl  // point to the cached data
    cache struct {
        RuntimeStatsColl
    }
}

vars.cache.StatementContext = stmtctx.StatementContext{}  // reset data
vars.StmtCtx = &vars.cache.StatementContext // allocate from the cached data

vars.StmtCtx.RuntimeStatsColl = &vars.StmtCtx.cache.RuntimeStatsColl  // also alloc from cached data, no extra object allocation!
```

It waste a pointer size MEM, but I think that's worthy.

Reply:
> One thing I noticed is that, &xxx{} will init all fields into the default zero value.

No, all the fields are initialized to default zero value, which is done in InitStatementContext()
https://github.com/pingcap/tidb/pull/26007/files#diff-db5df2c0ea43fa82bdc165fed66584b4c5ba481fb38cf7862889fc8afa64f081R871

@xhebox

Reply:
I think the following code do the same thing without adding a new field:

```
savedRuntimeStatsColl := vars.StmtCtx.RuntimeStatsColl
*vars.StmtCtx = stmtctx.StatementContext{}
vars.StmtCtx.RuntimeStatsColl = ResetRuntimeStatsColl(savedRuntimeStatsColl)
```

And I wonder the following code may have a chance not to allocate `savedSC` on heap:

```
savedSC := *vars.StmtCtx
.... // reset things and use the old SC
```

Reply:
Yes, you're right. @xhebox